omod/src/main/java/org/openmrs/module/pihcore/PihBasicAuthenticationScheme.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
+ * <p>
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations
+ * under the License.
+ * <p>
+ * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+ */
+package org.openmrs.module.pihcore;
+
+import org.apache.commons.lang.StringUtils;
+import org.openmrs.Location;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.context.ContextAuthenticationException;
+import org.openmrs.module.authentication.AuthenticationUtil;
+import org.openmrs.module.authentication.web.AuthenticationSession;
+import org.openmrs.module.authentication.web.BasicWebAuthenticationScheme;
+import org.openmrs.module.pihcore.metadata.Metadata;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.List;
+import java.util.Properties;
+
+import static org.openmrs.module.appui.UiSessionContext.LOCATION_SESSION_ATTRIBUTE;
+import static org.openmrs.module.emr.EmrConstants.COOKIE_NAME_LAST_SESSION_LOCATION;
+
+/**
+ * This expands on the BasicWebAuthenticationScheme to also handle collecting, validating, and setting sessionLocation
+ */
```

Comment:
This class extends the Basic username/password authentication scheme provided by the authentication module, such that it also handles the selection of session location during initial basic authentication.